src/main/java/net/minecraftforge/datafix/ForgeDataFixerDelegateHandler.java

```diff
@@ -0,0 +1,244 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * A delegate handler which can rebuild a data fixer on the fly.
+ */
+class ForgeDataFixerDelegateHandler extends DataFixerUpper
+{
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    private final int                              dataVersion;
+    private final List<ForgeSchema>                allSchemas;
+    private final Int2ObjectSortedMap<ForgeSchema> schemas;
+    private final List<ForgeDataFixDelegate> vanillaFixers;
+    private final IntSortedSet               vanillaFixerVersions;
+
+    private InnerFixer activeFixer;
+
+    ForgeDataFixerDelegateHandler(
+      final int dataVersion,
+      final Int2ObjectSortedMap<ForgeSchema> schemas,
+      final List<ForgeSchema> allSchemas,
+      final List<ForgeDataFixDelegate> vanillaFixers,
+      final Executor executor,
+      final IntSortedSet vanillaFixerVersions
+    )
+    {
+        //Supply dummy data to the super constructor
+        //We never invoke and method on the super and so this data does not matter.
+        super(Int2ObjectSortedMaps.emptyMap(), Collections.emptyList(), IntSortedSets.EMPTY_SET);
+
+        //Store the builders input so we can rebuild it later if needed.
+        this.dataVersion = dataVersion;
+        this.allSchemas = Collections.unmodifiableList(allSchemas);
+        this.schemas = schemas;
+        this.vanillaFixers = vanillaFixers;
+        this.vanillaFixerVersions = vanillaFixerVersions;
+
+        //Rebuild the setup immediately, don't trigger the lazy rebuild, it causes NPE's.
+        rebuildFixer(executor, false);
+    }
+
+    public void rebuildFixer(final Executor executor, final boolean triggerOffThreadCompute)
+    {
+        //Rebuild the total environment.
+        //Collect all fixes.
+        final List<ForgeDataFixDelegate> globalList = new ArrayList<>(this.vanillaFixers);
+
+        //Collect all schema version keys with known fixers.
+        final IntSortedSet fixerVersions = new IntAVLTreeSet(this.vanillaFixerVersions);
+
+        //And now collect all custom fixers again that we know of and update.
+        getAllSchemas().forEach(schema -> {
+            schema.getCustomFixers().forEach(fix -> {
+                final int version = DataFixUtils.getVersion(fix.getVersionKey());
+
+                if (version > dataVersion)
+                {
+                    LOGGER.warn("Ignored fix registered for version: {} as the DataVersion of the game is: {}", version, dataVersion);
+                    return;
+                }
+
+                final ForgeDataFixDelegate delegate = new ForgeDataFixDelegate(fix);
+
+                globalList.add(delegate);
+                fixerVersions.add(fix.getVersionKey());
+            });
+          });
+
+        //Reset our rules, normal vanilla rules (ours too) keep a cache of their schemas
```

Comment:
Just as clarification, does this comment say that all rules -- both vanilla and ours -- keep their schemas? Because if so, we can reword it to be more simple:
```suggestion
        //Reset our rules, rules keep a cache of their schemas
```

Reply:
What is not to understand.
All rules keep a cache of their schema.....
Not just ours.

Reply:
Fixed.